cpp/DEVELOPER_GUIDE.md

```diff
@@ -9,6 +9,84 @@ cuML is thread safe so its functions can be called from multiple host threads if
+3. Pointers to POD types (explicitly putting it out, even though can be considered as a POD).
+Internally, for the C++ base layer at least, these stateless functions are free to use their own temporary classes, as long as they are not exposed on the interface.
+
+### `libcuml++.so` (aka C++ API)
+Using the Decision Tree Classifier algorithm as an example, the following way of exposing its API would be wrong according to the guidelines in this section, since it exposes a non-POD C++ class object in the C++ API:
+```cpp
+template <typename T>
+class DecisionTreeClassifier {
+  TreeNode<T>* root;
+  DTParams params;
+  const cumlHandle &handle;
+public:
+  DecisionTreeClassifier(const cumlHandle &handle, DTParams& params, bool verbose=false);
+  void fit(const T *input, int n_rows, int n_cols, const int *labels);
+  void predict(const T *input, int n_rows, int n_cols, int *predictions);
+};
+
+void decisionTreeClassifierFit(const cumlHandle &handle, const float *input, int n_rows, int n_cols,
+                               const int *labels, DecisionTreeClassifier<float> *model, DTParams params,
+                               bool verbose=false);
+void decisionTreeClassifierPredict(const cumlHandle &handle, const float* input,
+                                   DecisionTreeClassifier<float> *model, int n_rows,
+                                   int n_cols, int* predictions, bool verbose=false);
+```
+
+An alternative correct way to expose this could be:
+```cpp
+// NOTE: this example assumes that TreeNode and DTParams are the model/state that need to be stored
+// and passed between fit and predict methods
+template <typename T> struct TreeNode { /* nested tree-like data structure, but written as a POD! */ };
+struct DTParams { /* hyper-params for building DT */ };
+typedef TreeNode<float> TreeNodeF;
+typedef TreeNode<double> TreeNodeD;
+
+void decisionTreeClassifierFit(const cumlHandle &handle, const float *input, int n_rows, int n_cols,
+                               const int *labels, TreeNodeF *&root, DTParams params,
+                               bool verbose=false);
+void decisionTreeClassifierPredict(const cumlHandle &handle, const double* input, int n_rows,
+                                   int n_cols, const TreeNodeD *root, int* predictions,
+                                   bool verbose=false);
+```
+The above example understates the complexity involved with exposing a tree-like data structure across the interface! However, this example should be simple enough to drive the point across.
+
+### WIP TODO (Need to explore this further!)
+This guidelines also mean that it is the responsibility of C++ API to expose methods to load and store (aka marshalling) such a data structure. Further continuing the Random Forest example,  the following methods could achieve this:
+```cpp
+void storeTreeNode(const TreeNodeF *root, std::ostream &os);
```

Comment:
Wondering if the naming of the function is as intended. Does `storeTreeNode` store the whole tree starting at `root`? If so I think it should be named `storeTree`.

Reply:
Done